tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,681 @@
+  // reference to op name string literals and c10 op registration/invocation
+  // APIs are almost always in the same function.
+  static void scanConnectedNodes(
+      Value* src,
+      const VALUE_SET& blocked,
+      const std::function<void(Value*)>& CB, LINK* debugLink) {
+    std::deque<Value*> worklist;
+    SmallPtrSet<Value*, 16> visited;
+
+    auto insert = [&](Value* cur, Value* parent) -> void {
+      if (!blocked.count(cur) && visited.insert(cur).second) {
+        worklist.push_back(cur);
+        if (debugLink) (*debugLink).emplace(cur, parent);
+      }
+    };
+
+    auto expandOperands = [&](Value* V) -> void {
+      // Stops if it doesn't have operands (!isa<User>) or it is a function.
+      if (!isa<User>(V) || isa<Function>(V)) return;
+      auto node = dyn_cast<User>(V);
+      for (auto& O : node->operands()) {
+        insert(O, node);
+      }
+    };
+
+    auto expandUsers = [&](Value* V) -> void {
+      // If the value is not constant, then the user of the value might pass
+      // other value into it, e.g.:
+      //   store @.str.15, %10
+      //   invoke @c10.reg_op, %10, @foo
+      // The store instruction, which is the user of "%10", passes "@.str.15" to
+      // "%10" which in turn is passed to "@c10.reg_op" API function.
+      // Users of constants are not interesting as they cannot change the state
+      // of the constant. We skip users of functions as well assuming
+      // interesting values (op names and function pointers) are not set via
+      // other invocations of the function.
+      if (!isa<User>(V) || isa<Constant>(V) || isa<Function>(V)) return;
+      for (auto U : V->users()) {
+        insert(U, V);
+      }
+    };
+
+    auto expand = [&](Value* V) -> void {
+      expandOperands(V);
+      expandUsers(V);
+    };
+
+    expand(src);
+    while (!worklist.empty()) {
+      auto cur = worklist.front();
+      worklist.pop_front();
+      expand(cur);
+
+      if (isa<Function>(cur) || isa<Constant>(cur)) {
+        CB(cur);
+      }
+    }
+  }
+
+  // Calculate transitive closure and remove intermediate (non-key) nodes.
+  // Note that there are two type of nodes in the dependency graph:
+  // 1) String literals in source files, e.g.:
+  //    "aten::cos_(Tensor(a!) self) -> Tensor(a!)", which represents operator
+  //    "schema";
+  // 2) Function symbols in object files, e.g.:
+  //    "at::CPUType::(anonymous namespace)::cos_(at::Tensor&)";
+  // Both of them are added to the dependency graph as std::string. Ultimately
+  // we only care about #1 as that's what we use to prune registered ops via
+  // codegen, then #2 will be stripped by linker automatically. So the goal is
+  // to remove #2 from the graph while maintaining the transitive dependency
+  // between #1. #1 is called "key nodes" in this method.
+  static void simplifyGraph(
+      GRAPH& input, SET& keyNodes, GRAPH* output, PATH* path) {
```

Comment:
It's just some strange habit I got long time ago :)
https://google.github.io/styleguide/cppguide.html#Output_Parameters

I'll try to add const to the input references; part of the reason I didn't do so because I was lazy to handle map[non-exist-key] case (inserting an empty value is fine for our use case).